### PR TITLE
3444 Add DatasetFunc

### LINKED ISSUE #3444: Add DecathlonDatalist to enhance decathlon style loading
**Is your feature request related to a problem? Please describe.**
Currently, we already have several useful utilities to load or sample the decathlon datalist, would be nice to add a class to combine and enhance the feature.
We already have a reference implementation in NVIDIA Clara and can config it in the JSON file:
```
{"datalist": {"name": "DecathlonDatalist", "args": {XXX}}},
{"dataset": {"name": "CacheDataset", "args": {"data": "@datalist"}}}
```
I will try to reimplement it according to MONAI style.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #3444 .

### Description
Thanks for @wyli 's sharing, this PR added the `DatasetFunc` referring to:
https://github.com/webdataset/webdataset/blob/master/notebooks/howitworks.ipynb

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/data/dataset.py [hunk 1]
```diff
@@ -97,6 +97,57 @@ def __getitem__(self, index: Union[int, slice, Sequence[int]]):
...
+        Reset the dataset items with specified `func`.
+
+        Args:
+            data: if not None, execute `func` on it, default to `self.src`.
+            func: if not None, execute the `func` with specified `kwargs`, default to `self.func`.
+            kwargs: other arguments for the `func` except for the first arg.
+
+        """
+        src = self.src if data is None else data
+        self.data = self.func(src, **self.kwargs) if func is None else func(src, **kwargs)
```
Comment (wyli) [inline monai/data/dataset.py:146]: This "self" and the returned "self.data" is used to set "self.data" in super(). This logic should be simplified

Comment (Nic-Ma) [reply monai/data/dataset.py:147]: Sure, simplified in the latest commit.

Thanks.

File: monai/data/dataset.py [hunk 2]
```diff
@@ -97,6 +97,58 @@ def __getitem__(self, index: Union[int, slice, Sequence[int]]):
...
+
+        return self.data
```
Comment (wyli) [inline monai/data/dataset.py:149]: Looks like now this return is not in use and could be removed.

Comment (Nic-Ma) [reply monai/data/dataset.py]: Make sense, removed `return` in the latest commit.

Thanks.

**Pull request merged**